PalasoUIWindowsForms/Reporting/MemoryManagement.cs

```diff
@@ -0,0 +1,91 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.Linq;
+using System.Reflection;
+using System.Text;
+using System.Text.RegularExpressions;
+using System.Windows.Forms;
+using L10NSharp;
+using Microsoft.VisualBasic.Devices;
+using Palaso.PlatformUtilities;
+using Palaso.Reporting;
+
+namespace Palaso.UI.WindowsForms.Reporting
+{
+	/// <summary>
+	/// This class contains methods useful for logging and warning the user about possible problems
+	/// of memory shortage.
+	/// </summary>
+	public static class MemoryManagement
+	{
+		private static bool s_warningShown;
+
+		/// <summary>
+		/// Write to the logger a message containing the event string, the total memory this process is using,
+		/// its working set (real memory), .NET heap memory, and the total physical and virtual memory on the system
+		/// If botherUser is true and total memory in use is more than 1G, it will display a dialog saying
+		/// "This program is using a lot of memory, and may soon slow down or experience other problems.
+		/// We recommend that you quit and restart it when convenient." This will happen only once per session.
+		/// Enhance: (Do we need a way to turn it off altogether? Should we have an argument whereby the threshold can be controlled?)
+		/// </summary>
+		/// <param name="minor">True to write a minor event, false to write full event</param>
+		/// <param name="eventDescription"></param>
+		/// <param name="okToBotherUser">The first time it is called with this true and short of memory, display a dialog.</param>
+		/// <returns>true if total memory in use is more than 1G (whether or not we bothered the user)</returns>
+		public static bool CheckMemory(bool minor, string eventDescription, bool okToBotherUser)
+		{
+			var heapMem = GC.GetTotalMemory(true); // first, as it may reduce other numbers
+			long memorySize64;
+			string message;
+			ulong totalPhysicalMemory = 0;
+			string totalVirtualMemory = "unknown";
+			if (Platform.IsWindows)
+			{
+				var computerInfo = new Computer().Info;
+				totalPhysicalMemory = computerInfo.TotalPhysicalMemory;
+				totalVirtualMemory = (computerInfo.TotalVirtualMemory / 1024).ToString("N0");
+			}
+			else if (Platform.IsLinux)
+			{
+				var meminfo = File.ReadAllText("/proc/meminfo");
+				var match = new Regex(@"MemTotal:\s+(\d+) kB").Match(meminfo);
+				if (match.Success)
+				{
+					totalPhysicalMemory = ulong.Parse(match.Groups[1].Value) * 1024;
+				}
+				// So far we have no way to get anything corresponding to Windows's idea of the total virtual memory
+				// the process can use.
```

Comment:
There are a couple of other lines in /proc/meminfo that may be of interest for getting this value:
SwapTotal:      16729084 kB
and
VmallocTotal:   34359738367 kB
SwapTotal is the total amount of physical swap memory (on disk).  This plus the MemTotal value might be what you want since I think it's the limit of what can actually be allocated by the kernel.
VmallocTotal gives the theoretical virtual memory size as far as I can tell from the sketchy descriptions I've seen.  (I could be wrong.)  Since addresses don't have to be contiguous, theoretically any address in that area could be allocated, but probably very few systems (if any) would be set up to allow all of those addresses to be allocated.  (if that's even possible for any CPU chip in production.)